lib/rules/no-uninstalled-addons.ts

```diff
@@ -0,0 +1,214 @@
+import { Property, ArrayExpression } from '@typescript-eslint/types/dist/ast-spec'
+
+//------------------------------------------------------------------------------
+// Rule Definition
+//------------------------------------------------------------------------------
+
+export = createStorybookRule({
+  name: 'no-uninstalled-addons',
+  defaultOptions: [],
+  meta: {
+    type: 'problem',
+    docs: {
+      description:
+        'This rule identifies storybook addons that are invalid because they are either not installed or contain a typo in their name.',
+      categories: [CategoryId.RECOMMENDED],
+      recommended: 'error', // or 'error'
+    },
+    messages: {
+      addonIsNotInstalled: `The {{ addonName }} is not installed. Did you forget to install it?`,
+    },
+
+    schema: [], // Add a schema if the rule has options. Otherwise remove this
+  },
+
+  create(context) {
+    // variables should be defined here
+
+    //----------------------------------------------------------------------
+    // Helpers
+    //----------------------------------------------------------------------
+
+    // this will not only exclude the nullables but it will also exclude the type undefined from them, so that TS does not complain
+    function excludeNullable<T>(item: T | undefined): item is T {
+      return !!item
+    }
+
+    type MergeDepsWithDevDeps = (packageJson: Record<string, string>) => string[]
+    const mergeDepsWithDevDeps: MergeDepsWithDevDeps = (packageJson) => {
+      const deps = Object.keys(packageJson.dependencies || {})
+      const devDeps = Object.keys(packageJson.devDependencies || {})
+      return [...deps, ...devDeps]
+    }
+
+    type IsAddonInstalled = (addon: string, installedAddons: string[]) => boolean
+    const isAddonInstalled: IsAddonInstalled = (addon, installedAddons) => {
+      // cleanup /register or /preset from registered addon
+      const addonName = addon.replace(/\/register$/, '').replace(/\/preset$/, '')
+      return installedAddons.includes(addonName)
+    }
+
+    type AreThereAddonsNotInstalled = (
+      addons: string[],
+      installedSbAddons: string[]
+    ) => false | { name: string }[]
+    const areThereAddonsNotInstalled: AreThereAddonsNotInstalled = (addons, installedSbAddons) => {
+      const result = addons
+        .filter((addon) => !isAddonInstalled(addon, installedSbAddons))
+        .map((addon) => ({ name: addon }))
+      return result.length ? result : false
+    }
+
+    type GetPackageJson = (path: string) => Record<string, any>
+
+    const getPackageJson: GetPackageJson = (path) => {
+      const packageJson = {
+        devDependencies: {},
+        dependencies: {},
+      }
+      try {
+        const file = readFileSync(path, 'utf8')
+        const parsedFile = JSON.parse(file)
+        packageJson.dependencies = parsedFile.dependencies || {}
+        packageJson.devDependencies = parsedFile.devDependencies || {}
+      } catch (e) {
+        throw new Error(
```

Comment:
This relies too much on package.json being in the exact same folder, which is not always the case. For instance in our projects we tend to put the config files for storybook into its own folder inside `config/storybook` while the project `package.json` remains in the top level folder. Is there a chance to provide a configuration for this or extend the code so it drills up through the folders like npm looks like to do?

Reply:
Hey @jdehaan thanks for this comment! Would something like this suffice?

```json
{
  "overrides": [
    {
      "files": ["whatever-is-your-config-dir/main.@(js|ts)"],
      "rules": {
        "storybook/no-uninstalled-addons": ["error", { "packageLocation": "./" }],
      }
    }
  ]
}
```

1 - You are able to set the location of the main.js file which will get the rule applied (currently possible)
2 - You are able to set the location of which the package.json that contains the addons is located at (needs to be implemented)


----


Another possible solution would be to use something like [`pkgUp`](https://github.com/NicolasSchwarzer/pkgup) to find the closest package.json, which could work in your scenario, but still fail in scenarios where the package.json that was found is not the correct one.

Reply:
Thanks for the quick feedback.

I also think that an explicit option is the safest way to go, these things won't jump up and down at all times and a one time configuration effort is fully ok and preferable to some black magic IMO. For very fancy configuration needs one could even provide an own configuration logic upfront generating pairs of main.js/package.json for bigger monorepos... So yes I would be fully happy with your configuration proposal!

Reply:
@yannbf Currently, if `package.json` is not found, the error is only [logged](https://github.com/storybookjs/eslint-plugin-storybook/blob/fa6290d5656d0184a911c5eeae88e1051f8db027/lib/rules/no-uninstalled-addons.ts#L147) to the console. It should probably be re-thrown or handled in a way that results in a non-zero exit status. At the moment, the linting process will silently pass.

Reply:
Hey @jdehaan @menelaos, a new version has been released including this feature! https://github.com/storybookjs/eslint-plugin-storybook/pull/102

And @menelaos that was intentional. We don't want people's projects to break so this plugin blocks them. The plugin now just acknowledges the issue and provides a link to the docs on how to fix it.

Thank you all!

Reply:
Works like a charm, beside that the new package features a smarter package.json resolution as I did not have to specify it anymore in many cases. Thanks!

Reply:
All thanks to @andrelas1 great work!

Reply:
@yannbf @andrelas1 Thank you!